evalml/tests/automl_tests/test_engine_base.py

```diff
@@ -42,3 +45,59 @@ def test_train_and_score_pipelines_error(mock_fit, mock_score, dummy_binary_pipe
     for i in range(automl.data_splitter.get_n_splits()):
         assert np.isnan(evaluation_result['cv_data'][i]['all_objective_scores']['Log Loss Binary'])
     assert 'yeet' in caplog.text
+
+
+@patch('evalml.objectives.BinaryClassificationObjective.optimize_threshold')
+@patch('evalml.pipelines.BinaryClassificationPipeline.predict_proba')
+@patch('evalml.pipelines.BinaryClassificationPipeline.fit')
+@patch('evalml.automl.engine.engine_base.split_data')
+def test_train_pipeline_trains_and_tunes_threshold(mock_split_data, mock_pipeline_fit,
+                                                   mock_predict_proba, mock_optimize, X_y_binary,
+                                                   dummy_binary_pipeline_class):
+    X, y = X_y_binary
+    mock_split_data.return_value = split_data(X, y, "binary", test_size=0.2, random_seed=0)
+
+    _ = EngineBase.train_pipeline(dummy_binary_pipeline_class({}), X, y,
+                                  optimize_thresholds=True, objective=LogLossBinary())
+
+    mock_pipeline_fit.assert_called_once()
+    mock_optimize.assert_not_called()
+    mock_split_data.assert_not_called()
+
+    mock_pipeline_fit.reset_mock()
+    mock_optimize.reset_mock()
+    mock_split_data.reset_mock()
+
+    _ = EngineBase.train_pipeline(dummy_binary_pipeline_class({}), X, y,
+                                  optimize_thresholds=True, objective=F1())
+    mock_pipeline_fit.assert_called_once()
+    mock_optimize.assert_called_once()
+    mock_split_data.assert_called_once()
+
+
+def test_engines_check_pipeline_names_unique(dummy_binary_pipeline_class):
+
+    class MinimalEngine(EngineBase):
+
+        def evaluate_batch(self, pipelines):
+            """Docstring so codecov doesn't complain."""
+
+        def train_batch(self, pipelines):
+            super().train_batch(pipelines)
+
+        def score_batch(self, pipelines, X, y, objectives):
+            super().score_batch(pipelines, X, y, objectives)
+
+    engine = MinimalEngine(X_train=pd.DataFrame(), y_train=pd.Series())
+
+    class Pipeline1(dummy_binary_pipeline_class):
+        custom_name = "My Pipeline"
+
+    class Pipeline2(dummy_binary_pipeline_class):
+        custom_name = "My Pipeline"
+
+    with pytest.raises(ValueError, match="All pipeline names must be unique. The names My Pipeline were repeated."):
+        engine.train_batch([Pipeline2({}), Pipeline1({})])
+
+    with pytest.raises(ValueError, match="All pipeline names must be unique. The names My Pipeline were repeated."):
+        engine.score_batch([Pipeline2({}), Pipeline1({})], None, None, None)
```

Comment:
You mean if you passed in a list of `None` as the pipelines. Yea that would raise an error.

I used `None` for the `X`, `y` and `objectives` just because we don't need "real" values to test whether the unique name assertion error is raised. I can use real values if you think that's better!

Reply:
Ah, completely read that wrong! My bad, this looks fine then